.github/actions/test-setup/action.yml

```
@@ -0,0 +1,38 @@
+name: "Setup Python and Virtual Environment"
+description: "Sets up Python, a virtual environment, and installs dependencies"
+
+inputs:
+  python-version:
+    description: "Python version to use"
+    required: true
+  venv-dir:
+    description: "Virtual environment directory"
+    required: true
+    default: "${{ github.workspace }}/.venv"
+
+runs:
+  using: "composite"
+  steps:
+    - name: Get Current Date
+      id: date
+      shell: bash
+      run: echo "date=$(date +'%Y-%m-%d')" >> $GITHUB_ENV
+
+    - name: Set up Python
+      uses: actions/setup-python@v5.4.0
```

```suggestion
      uses: actions/setup-python@42375524e23c412d93fb67b49958b491fce71c38 # v5.4.0
```
We're pinned with hashes as suggested by OpenSSF Scorecard, see https://github.com/ossf/scorecard/blob/main/docs/checks.md#pinned-dependencies.